File: two-sum/TwoSum.js
function twoNumberSum1(array, targetSum) {
    for(let i = 0; i < array.length; i++) {
      for(let j = 0; j < array.length; j++) {
        if(i === j) continue;
        if(array[i] + array[j] === targetSum) { return [ array[i], array[j] ]; }
      }
    }
    return [];
  }

function twoNumberSum2(array, targetSum) {
  for(let i = 0; i < array.length; i++) {
    for(let j = i + 1; j < array.length; j++) {
      if(array[j] + array[i] === targetSum) {
        return [ array[i], array[j] ];
      }
    }
  }
  return [];
}

function twoNumberSum3(array, targetSum) {
    const set = new Set();
    for(const num of array) {
      let difference = targetSum - num;
      if(set.has(difference)) { return [ num, difference ]; }
      set.add(num)
    }
    return [];
}

function twoNumberSum4(array, targetSum) {
    const map = new Map();
    for(const num of array) {
      let difference = targetSum - num;
      if(map.has(difference)) {
        return [ difference, map.get(difference) ]
      }
      map.set(num, difference)
    }
    return [];
}

function twoNumberSum(array, targetSum) {
    array.sort((a, b) => a - b);
    var left = 0;
    var right = array.length - 1;
    while(left < right) {
      var sum = array[left] + array[right];
      if(sum == targetSum) {
        return [ array[left], array[right] ];
      } else if (sum < targetSum) {
        left++;
      } else {
        right--;
      }
    }
    return [];
}

exports.twoNumberSum = twoNumberSum;

File: two-sum/TwoSum.cs
using System;
using System.Collections.Generic;

public class TwoSum {
    public static int[] TwoNumberSum1(int[] array, int targetSum) {
        for(int i = 0; i < array.Length; i++) {
            for(int j = 0; j < array.Length; j++) {
                if(i == j) continue; 
                if(array[i] + array[j] == targetSum) { 
                    return new int[] { array[i], array[j] }; 
                }
            }
        }
        return new int[0];                
    }

    public static int[] TwoNumberSum2(int[] array, int targetSum) {
        for(int i = 0; i < array.Length; i++) {
            for(int j = i + 1; j < array.Length; j++) {
                if(array[i] + array[j] == targetSum) { 
                    return new int[] { array[i], array[j] }; 
                }
            }
        }
        return new int[0];
    }

    public static int[] TwoNumberSum3(int[] array, int targetSum) {
        HashSet<int> set = new HashSet<int>();
        int index = 0;
        while(index < array.Length) {
            int difference = targetSum - array[index];
            if(set.Contains(difference)) { return new int[] { array[index], difference }; }
            set.Add(array[index]);
            index++;
        }
        return new int[]{};
    }

    public static int[] TwoNumberSum4(int[] array, int targetSum) {
        Dictionary<int, int> dictionary = new Dictionary<int, int>();
        foreach(int num in array) {
            int difference = targetSum - num;
            if(dictionary.ContainsKey(difference)) {
                return new int[] { difference, num };
            }
            dictionary.Add(num, difference);
        }
        return new int[]{};
    }

    public static int[] TwoNumberSum5(int[] array, int targetSum) {
        Array.Sort(array);
        int left = 0;
        int right = array.Length - 1;
        while(left < right) {
            int sum = array[left] + array[right];
            if(sum == targetSum) {
                return new int[]{ array[left], array[right] };
            } else if (sum < targetSum) {
                left++;
            } else {
                right--;
            }
        }
        return new int[]{};
    }
}